packages/bff/src/services/purposeService.ts

```diff
@@ -4,16 +4,277 @@ import {
+    eserviceClient: CatalogProcessClient;
+    tenantClient: TenantProcessClient;
+    agreementClient: AgreementProcessClient;
+    authorizationClient: AuthorizationProcessClient;
+  },
+  requesterId: string,
+  filters: {
+    name?: string | undefined;
+    eservicesIds?: string[];
+    consumersIds?: string[];
+    producersIds?: string[];
+    states?: purposeApi.PurposeVersionState[];
+    excludeDraft?: boolean | undefined;
+  },
+  offset: number,
+  limit: number,
+  headers: Headers
+): Promise<bffApi.Purposes> {
+  const purposes = await purposeClient.getPurposes({
+    queries: {
+      ...filters,
+      limit,
+      offset,
+    },
+    withCredentials: true,
+    headers,
+  });
+
+  const eservices = await Promise.all(
+    [...new Set(purposes.results.map((p) => p.eserviceId))].map((id) =>
+      eserviceClient.getEServiceById({
+        params: {
+          eServiceId: id,
+        },
+        withCredentials: true,
+        headers,
+      })
+    )
+  );
+
+  const getTenant = async (id: string): Promise<tenantApi.Tenant> =>
+    tenantClient.tenant.getTenant({
+      params: {
+        id,
+      },
+      withCredentials: true,
+      headers,
+    });
+  const consumers = await Promise.all(
+    [...new Set(purposes.results.map((p) => p.consumerId))].map(getTenant)
+  );
+  const producers = await Promise.all(
+    [...new Set(eservices.map((e) => e.producerId))].map(getTenant)
+  );
+
+  const enhancePurpose = async (
+    purpose: purposeApi.Purpose
+  ): Promise<bffApi.Purpose> => {
+    const eservice = eservices.find((e) => e.id === purpose.eserviceId);
+    if (!eservice) {
+      throw eServiceNotFound(unsafeBrandId(purpose.eserviceId));
+    }
+
+    const producer = producers.find((p) => p.id === eservice.producerId);
+    if (!producer) {
+      throw tenantNotFound(unsafeBrandId(eservice.producerId));
+    }
+
+    const consumer = consumers.find((c) => c.id === purpose.consumerId);
+    if (!consumer) {
+      throw tenantNotFound(unsafeBrandId(purpose.consumerId));
+    }
+
+    const latestAgreement = await getLatestAgreement(
+      agreementClient,
+      purpose.consumerId,
+      eservice,
+      headers
+    );
+    if (!latestAgreement) {
+      throw agreementNotFound(unsafeBrandId(purpose.consumerId));
+    }
+
+    const currentDescriptor = eservice.descriptors.find(
+      (d) => d.id === latestAgreement.descriptorId
+    );
+    if (!currentDescriptor) {
+      throw eServiceDescriptorNotFound(
+        unsafeBrandId(eservice.id),
+        unsafeBrandId(latestAgreement.descriptorId)
+      );
+    }
+
+    const clients =
+      requesterId === purpose.consumerId
+        ? (
+            await getAllClients(
+              authorizationClient,
+              purpose.consumerId,
+              purpose.id,
+              headers
+            )
+          ).map((c) => ({
+            id: c.client.id,
+            name: c.client.name,
+            hasKeys: c.keys.length > 0,
+          }))
```

Comment:
Cherry pick and reuse this converter created by @paologaleotti ? https://github.com/pagopa/interop-be-monorepo/pull/753/files#diff-6113775e4e2b97fdc29f45864600186a420e532ae9f67e30b880c26b5a8f056dR101